docs/source/tutorials.md

```diff
@@ -101,3 +101,197 @@ Since the target repository is different than the source, it has to be specified
+        then if (alpha + beta + gamma == 180 )
+            then Just (MkTriangle alpha beta gamma) 
+            else Nothing
+        else Nothing
+
+{-# COMPILE AGDA2HS createTriangle #-}
+```
+Unfortunately, the instance arguments are inferred only if presented in exactly the same shape as hypotheses in the constructor, which is the cause of perceived redundancy of the if-then-else statement - chaining the hypotheses with `&&` (the AND operator defined in `Haskell.Prim.Bool` in Agda2hs) will not allow for the reference to be inferred automatically, and the proofs would have to be provided manually. An example for that can be seen in the alternative `createTriangle₁` function:
+
+```agda
+createTriangle₁ : Nat -> Nat -> Nat -> Maybe Triangle
+createTriangle₁ alpha beta gamma 
+    = if ((countBiggerThan (alpha ∷ beta ∷  gamma ∷ []) 90 <= 1) && (alpha + beta + gamma == 180 )) 
+        then (λ ⦃ h₁ ⦄ →  Just (MkTriangle alpha beta gamma 
+            ⦃ &&-rightTrue (countBiggerThan (alpha ∷ beta ∷  gamma ∷ []) 90 <= 1) (alpha + beta + gamma == 180 ) h₁ ⦄ 
+            ⦃ &&-leftTrue (countBiggerThan (alpha ∷ beta ∷  gamma ∷ []) 90 <= 1) (alpha + beta + gamma == 180 ) h₁ ⦄) )
+        else Nothing
+ 
+{-# COMPILE AGDA2HS createTriangle₁ #-}
+
+```
+While using this alternative function offers much cleaner Haskell code as an output (no nested if statements), however the Agda side gets a bit messier. Two things are worth noting here: first, to be able to operate on the hypothesis asserted in the if condition, the branch has to be rewritten as an anonymous function taking the assertion as instance argument. Secondly, to explicitly provide the instance arguments, it also has to be done inside the double curly brackets. 
+
+Lastly, the two functions used to extract conditions from the compound condition: `&&-rightTrue` and `&&-leftTrue` were defined in `Haskell.Law.Bool`. The `Law` folder contains many theorems useful for designing your own proofs. This will be expanded on in following sections. 
+
+### Function or Predicate?
+
+The code described in this section can be found in the file `Ascending.agda`. 
+
+We will try to define ascending order on lists, which will allow us to use statements about the order in later proofs and programs. 
+
+A first attempt at definition could be a function that can provide judgments on instances of lists:
+
+```agda
+isAscending : ⦃ iOrdA : Ord a ⦄ → List a → Bool
+isAscending [] = True
+isAscending (x ∷ []) = True
+isAscending (x ∷ y ∷ xs) = if x <= y then isAscending (y ∷ xs) else False
+
+{-# COMPILE AGDA2HS isAscending #-}
+```
+This function can be compiled to Haskell without any issue, however, when you try using it in proofs you can notice that it is not the most handy definition: the different cases are anonymous and to invoke them, some heavy stretching with regards to values has to be invoked. A better definition might be a predicate, instead:
```

Comment:
> some heavy stretching with regards to values has to be invoked

I'm not sure what you mean by this, could you clarify?

Reply:
Is the new version better?